src/types.jl

```diff
@@ -0,0 +1,118 @@
+"""
+    Assignable
+
+`Assignable` represents any data type that can be associated with traits.
+It essentially covers all data types including parametric types e.g. `AbstractArray`
+"""
+const Assignable = Union{UnionAll, DataType}
+
+"""
+    Contract{T <: DataType, F <: Function, N}
+
+A contract refers to a function defintion `func` that is required to satisfy
+the Can-type of a trait. The function `func` must accepts `args` and returns `ret`.
+
+# Fields
+- `can_type`: can-type of a trait e.g. `CanFly`
+- `func`: function that must be implemented to satisfy this trait
+- `args`: argument types of the function `func`
+- `kwargs`: keyword argument names of the function `func`
+- `ret`: return type of the function `func`
+"""
+struct Contract{T <: DataType, F <: Function}
+    can_type::T
+    func::F
+    args::Tuple
+    kwargs::Tuple
+    ret::Union{DataType,Nothing}
+end
+
+function Base.show(io::IO, c::Contract)
+    typ = Symbol(TYPE_PLACEHOLDER)
+    args = string("(", join([typ, c.args...], ", ::"))
+    if length(c.kwargs) > 0
+        args = string(args, "; ", join(c.kwargs, ", "))
+    end
+    args = string(args, ")")
+    trait = supertype(c.can_type)
+    print(io, "$(trait): $(c.can_type) ⇢ $(c.func)$(args)")
+    c.ret !== nothing && print(io, "::$(c.ret)")
+end
+
+"""
+    InterfaceReview
+
+An InterfaceReview object contains the validation results of an interface.
+
+# Fields
+- `type`: the type being checked
+- `result`: true if the type fully implements all required contracts
+- `implemented`: an array of implemented contracts
+- `misses`: an array of unimplemented contracts
+"""
+@Base.kwdef struct InterfaceReview
+    type::Assignable
```

Comment:
Can you replace the field name `type` by something else. While it is syntactically correct to use this Julia keyword here, a more expressive name (like `assigned_type`) would be better. It also spoils my editor (vim), if `type` is used in certain contexts.

Reply:
Done.  I renamed to `data_type`.